pkg/graceful/graceful.go

```diff
@@ -0,0 +1,173 @@
+// Package graceful contains tools for graceful shutdown.
+// GS refers to the process of shutting down a system in a controlled manner, allowing it to complete ongoing tasks,
+// release resources, and perform necessary cleanup operations before terminating.
+// This ensures that the system stops functioning without causing data loss, corruption, or other issues.
+package graceful
+
+import (
+	"context"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+)
+
+// Process represents "virtual" process that contains
+// routines that can be started and stopped
+// Note that ANY failure in starting a service will cause the process to shutdown
+type Process struct {
+	stop      <-chan os.Signal
+	stopStack []func()
+
+	timeout time.Duration
+	mu      sync.Mutex
+	stopped bool
+
+	logger zerolog.Logger
+}
+
+// Service represents abstract service.
+type Service interface {
+	Start(ctx context.Context) error
+	Stop()
+}
+
+// DefaultProcess is a process instance with some sane defaults.
+var DefaultProcess = New(15*time.Second, log.Logger, NewSigChan(syscall.SIGINT, syscall.SIGTERM))
+
+// New Process constructor.
+func New(timeout time.Duration, logger zerolog.Logger, stop <-chan os.Signal) *Process {
+	return &Process{
+		stop:    stop,
+		timeout: timeout,
+		logger:  logger.With().Str("module", "graceful").Logger(),
+	}
+}
+
+// AddService adds Service to the process.
+func (p *Process) AddService(ctx context.Context, s Service) {
+	p.AddStarter(ctx, s.Start)
+	p.AddStopper(s.Stop)
+}
+
+// AddStarter runs a function that starts something.
+// This is a blocking call for blocking .Start() services
+func (p *Process) AddStarter(ctx context.Context, fn func(ctx context.Context) error) {
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				p.logger.Error().Interface("panic", r).Msg("panic in service")
+				p.ShutdownNow()
+			}
+		}()
```

Comment:
Please make sure to log the actual panic source like https://github.com/zeta-chain/node/pull/2881